cmd/openshift-tests/disruption.go

```diff
@@ -0,0 +1,175 @@
+	restConfig, err := monitor.GetMonitorRESTConfig()
+	if err != nil {
+		return err
+	}
+
+	lb := backend.ParseStringToLoadBalancerType(opt.LoadBalancerType)
+
+	ctx, cancelFn := context.WithCancel(context.Background())
+	defer cancelFn()
+	abortCh := make(chan os.Signal, 2)
+	go func() {
+		<-abortCh
+		fmt.Fprintf(opt.ErrOut, "Interrupted, terminating\n")
+		// Give some time to store intervals on disk
+		time.Sleep(5 * time.Second)
+		cancelFn()
+		sig := <-abortCh
+		fmt.Fprintf(opt.ErrOut, "Interrupted twice, exiting (%s)\n", sig)
+		switch sig {
+		case syscall.SIGINT:
+			os.Exit(130)
+		default:
+			os.Exit(0)
+		}
+	}()
+	signal.Notify(abortCh, syscall.SIGINT, syscall.SIGTERM)
+
+	m, err := StartAPIAvailability(ctx, restConfig, lb)
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		ticker := time.NewTicker(100 * time.Millisecond)
+		defer ticker.Stop()
+		var last time.Time
+		done := false
+		for !done {
+			select {
+			case <-ticker.C:
+			case <-ctx.Done():
+				done = true
+			}
+			events := m.Intervals(last, time.Time{})
+			if len(events) > 0 {
+				for _, event := range events {
+					if !event.From.Equal(event.To) {
+						continue
+					}
+					fmt.Fprintln(opt.Out, event.String())
+				}
+				last = events[len(events)-1].From
+			}
+		}
+	}()
+
+	<-ctx.Done()
+
+	// Store intervals to artifact directory
+	intervals := m.Intervals(time.Time{}, time.Time{})
+	if len(opt.ExtraMessage) > 0 {
+		fmt.Fprintf(opt.Out, "\nAppending %s to recorded event message\n", opt.ExtraMessage)
+		for i, event := range intervals {
+			intervals[i].Message = fmt.Sprintf("%s user-provided-message=%s", event.Message, opt.ExtraMessage)
+		}
+	}
+
+	eventDir := filepath.Join(opt.ArtifactDir, monitorapi.EventDir)
+	if err := os.MkdirAll(eventDir, os.ModePerm); err != nil {
+		fmt.Printf("Failed to create monitor-events directory, err: %v\n", err)
+		return err
+	}
+
+	timeSuffix := fmt.Sprintf("_%s", time.Now().UTC().Format("20060102-150405"))
+	if err := monitor.WriteEventsForJobRun(eventDir, nil, intervals, timeSuffix); err != nil {
+		fmt.Printf("Failed to write event data, err: %v\n", err)
+		return err
+	}
+	fmt.Fprintf(opt.Out, "\nEvent data written, exiting\n")
+
+	return nil
+}
+
+// StartAPIAvailability monitors just the cluster availability
+func StartAPIAvailability(ctx context.Context, restConfig *rest.Config, lb backend.LoadBalancerType) (*monitor.Monitor, error) {
+	m := monitor.NewMonitorWithInterval(time.Second)
+	client, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return nil, err
+	}
+	if err := controlplane.StartAPIMonitoringUsingNewBackend(ctx, m, restConfig, lb); err != nil {
+		return nil, err
+	}
+
+	// read the state of the cluster apiserver client access issues *before* any test (like upgrade) begins
+	intervals, err := apiserveravailability.APIServerAvailabilityIntervalsFromCluster(client, time.Time{}, time.Time{})
+	if err != nil {
+		klog.Errorf("error reading initial apiserver availability: %v", err)
```

Comment:
Is there a reason we only log the error but not return it? Is it OK to add a potential nil to the monitor?

Reply:
API availability may start as a sidecontainer to watch resources, before cluster is created. We should not throw errors if API is not available initially